cocos/platform/CCFileUtils.cpp

```diff
@@ -43,6 +43,44 @@ THE SOFTWARE.
 
 NS_CC_BEGIN
 
+SpinRWLock::SpinRWLock(): _state({false, 0}) {}
+
+void SpinRWLock::readLock() {
+    State newState {false, 0}, oldState = _state.load(std::memory_order_acquire);
+    do {
+        oldState.writeFlag = false; // if writing, then wait
+        newState.readCount = oldState.readCount + 1;
+    } while (!_state.compare_exchange_weak(oldState, newState, std::memory_order_acquire, std::memory_order_relaxed));
+}
+
+void SpinRWLock::readUnLock() {
+    State newState, oldState = _state.load(std::memory_order_acquire);
+    do {
+        newState = oldState;
+        --newState.readCount;
+    } while (!_state.compare_exchange_weak(oldState, newState, std::memory_order_release, std::memory_order_relaxed));
+}
+
+void SpinRWLock::writeLock() {
+    State newState {true, 0}, oldState = _state.load(std::memory_order_acquire);
+    // set writeFlag first to hold off incoming readers
+    do {
+        newState.readCount = oldState.readCount;
+        oldState.writeFlag = false; // if writing, then wait
+    } while (!_state.compare_exchange_weak(oldState, newState, std::memory_order_acquire, std::memory_order_relaxed));
+    // wait for all ongoing readers
+    do {
+        newState = atomic_load_explicit(&_state, std::memory_order_acquire);
+    } while (newState.readCount);
+}
+
+void SpinRWLock::writeUnlock() {
+    State newState {false, 0}, oldState = _state.load(std::memory_order_acquire);
+    do {
+        newState.readCount = oldState.readCount;
+    } while (!_state.compare_exchange_weak(oldState, newState, std::memory_order_release, std::memory_order_relaxed));
+}
+
```

Comment:
Yeah, but I'm not sure which is the best place to put the file so I'd be expecting some comments from the maintainers.